internal/balancer/balanceraggregator/balanceraggregator.go

```diff
@@ -0,0 +1,168 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package balanceraggregator implements a BalancerAggregator helper.
+package balanceraggregator
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/internal/balancer/gracefulswitch"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+// ChildState is the balancer state of the child along with the
+// endpoint which ID's the child balancer.
+type ChildState struct {
+	Endpoint resolver.Endpoint
+	State    balancer.State
+}
+
+// Parent is a balancer.ClientConn that can also receive
+// child state updates.
+type Parent interface {
+	balancer.ClientConn
+	UpdateChildState(childStates []ChildState)
+}
+
+// Build returns a new BalancerAggregator.
+func Build(parent Parent, opts balancer.BuildOptions) *BalancerAggregator {
+	return &BalancerAggregator{
+		parent:   parent,
+		bOpts:    opts,
+		children: resolver.NewEndpointMap(),
+	}
+}
+
+// BalancerAggregator is a balancer that wraps child balancers. It creates a
+// child balancer with child config for every Endpoint received. It updates the
+// child states on any update from parent or child.
+type BalancerAggregator struct {
+	parent Parent
+	bOpts  balancer.BuildOptions
+
+	children *resolver.EndpointMap
+
+	inhibitChildUpdates bool
+}
+
+func (ba *BalancerAggregator) UpdateClientConnState(state balancer.ClientConnState) error {
+	endpointSet := resolver.NewEndpointMap()
+	ba.inhibitChildUpdates = true
+	// Update/Create new children.
+	for _, endpoint := range state.ResolverState.Endpoints {
+		endpointSet.Set(endpoint, nil)
+		var bal *balancerWrapper
+		if child, ok := ba.children.Get(endpoint); ok {
+			bal = child.(*balancerWrapper)
+		} else {
+			bal = &balancerWrapper{
+				endpoint:   endpoint,
+				ClientConn: ba.parent,
+				ba:         ba,
+			}
+			bal.Balancer = gracefulswitch.NewBalancer(bal, ba.bOpts)
+			ba.children.Set(endpoint, bal)
+		}
+		if err := bal.UpdateClientConnState(balancer.ClientConnState{
+			BalancerConfig: state.BalancerConfig,
+			ResolverState: resolver.State{
+				Endpoints:  []resolver.Endpoint{endpoint},
+				Attributes: state.ResolverState.Attributes,
+			},
+		}); err != nil {
+			return fmt.Errorf("error updating child balancer: %v", err)
+		}
+	}
+	// Delete old children that are no longer present.
+	for _, e := range ba.children.Keys() {
+		child, _ := ba.children.Get(e)
+		bal := child.(balancer.Balancer)
+		if _, ok := endpointSet.Get(e); !ok {
+			bal.Close()
+			ba.children.Delete(e)
+		}
+	}
+	ba.inhibitChildUpdates = false
+	ba.updateChildStates()
```

Comment:
Maybe `defer` these (in `ResolverError` too)?

Otherwise note the bug that exists when an error occurs and you return early.

Reply:
Hmm right I don't set it to false. Added defer in both operations. Thinking about it though, this question applies to the endpoint map too. If I return early, I'm midway through adding (and haven't yet deleted the children that need to be cleaned up), and thus have this weird interspliced state where the children are partially updated and no longer viable ones no longer stick around. Any thoughts here?